### DISCUSSION TITLE
Delay of recordToFile/streamVideo after motion event

### DISCUSSION THREAD

**Discussion by matthias-straub** [dgreif/ring#1240] [Q&A] [answered]
Hi, 
first of all, thank you for this great project. The API worked like a breeze for me. 
When recording a video after a motion event, I have about a 10 second delay compared to the videos that Ring provides via their subscription service. I assume that Ring is able to ask the camera to go back in time a bit. Am I correct here, and is this something the API can provide as well?
Cheers,
Matthias

**Comment by dgreif** [comment]
It's likely that Ring is keeping a rolling buffer of a few seconds on the camera, and they can pull that buffer when motion is detected. I don't know of a way to access that buffer via the API, so the delay is likely not avoidable unfortunately.

**Reply by tsightler** [reply, marked as answer]
As far as I can tell there's basically three modes of operation, legacy cameras without pre-roll, which appear to record maybe a second or so before the motion event (perhaps just the buffer used for detecting motion), and then newer cameras with pre-roll/advanced pre-roll, with the pre-roll behavior varying between battery and line powered cameras.  Battery cameras appear to use a low-power, low-res, black-and-white sensor to record into a pre-roll buffer, while line powered camera appear to simply use a lower resolution mode for the existing sensor for pre-roll recordings.

In all cases, the camera is providing this recording and uploading them to Ring servers.  I'm not sure if they are combined on the device or after, but, based on my analysis of the recording files, at least for legacy cameras without pre-roll, it seems like it's just two separate H.264 files crudely stiched together which makes me think it might happen on the camera itself.  Regardless, none of that is available from the Ring app API since this API only provides access to the same functions used by the Ring app while the protocol and API used between Ring cameras and servers (Ring Socket Services) is largely opaque.

**Reply by matthias-straub** [reply]
Thank you for your replies!
Best regards
Matthias
